File: src/datastructures/DSU.java
package datastructures;

public class DSU {
    int[] parents;
    int[] size;

    DSU (int n) {
        this.parents = new int[n];
        this.size = new int[n];
        for (int i = 0; i < n; i++) {
            parents[i] = i;
            size[i] = 1;
        }
    }

    int find(int x) {
        if (parents[x] != x) parents[x] = find(parents[x]);
        return parents[x];
    }

    boolean union(int x, int y) {
        var xParent = find(x);
        var yParent = find(y);

        if (xParent == yParent) return false;

        if (size[xParent] > size[yParent]) {
            parents[yParent] = xParent;
            size[xParent] += size[yParent];
        } else {
            parents[xParent] = yParent;
            size[yParent] += size[xParent];
        }

        return true;
    }
}

File: templates/cpp/datastructures/DSU.cpp
#include <bits/stdc++.h>

using namespace std;

struct DSU {
    vector<int> parents;
    vector<int> rank;

    DSU(int size) {
        parents.assign(size, 0);
        for (int i = 0; i < size; i++) {
            parents[i] = i;
        }
        rank.assign(size, 1);
    }

    int find(int x) {
        if (parents[x] != x) {
            parents[x] = find(parents[x]);
        }
        return parents[x];
    }

    bool unite(int x, int y) {
        int xParent = find(x), yParent = find(y);
        if (xParent == yParent) return false;

        if (rank[xParent] > rank[yParent]) {
            parents[yParent] = xParent;
            rank[xParent] += rank[yParent];
        } else {
            parents[xParent] = yParent;
            rank[yParent] += rank[xParent];
        }
        return true;
    }

    bool areConnected(int x, int y) {
        return find(x) == find(y);
    }
};

int main() {
    // --- Test Case 1: Linear Chain (Basic Connectivity) ---
    DSU dsu(5);
    dsu.unite(0, 1);
    dsu.unite(1, 2);
    dsu.unite(3, 4);

    assert(dsu.areConnected(0, 2)); // 0 and 2 should be connected
    assert(!dsu.areConnected(0, 3)); // 0 and 3 should NOT be connected
    cout << "Test Case 1 Passed: Basic connectivity confirmed." << endl;
    return 0;
}